lib/tree.rb

```diff
@@ -2,61 +2,142 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(logn) because one side of the tree can be eliminated each time the new key is evaluated against the existing nodes data
```

Comment:
It will be O(log n) if the tree is balanced but O(n) if the tree is totally unbalanced.